### PR TITLE
Challenge 14 basic calculator

### PR DESCRIPTION
**Pull request by AnaiMurillo**
---
name: Challenge Basic Calculator
about: A template PR for code review with a checklist
---

## Behavior

Create a simple calculator program in Python that can perform basic arithmetic operations: addition, subtraction, multiplication, and division.

Addition: Adds two numbers together.
Subtraction: Subtracts the second number from the first number.
Multiplication: Multiplies two numbers.
Division: Divides the first number by the second number. Make sure to handle division by zero.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/basic_calculator/main.py
Comment (hectordacb) [inline solutions/basic_calculator/main.py:1]: This file does not have the correct name. 
[ ] The file name describes the function's behavior

File: solutions/tests/test_basic_calculator.py
Comment (hectordacb) [inline solutions/tests/test_basic_calculator.py:1]: Each test should only have one assertion

File: solutions/basic_calculator/main.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+Module contents:
+    - calculate: Performs the specified arithmetic operation on two numbers
+
+Created on 28/12/2024
+Author: Ana Isabel Murillo
+"""
+
+
+def calculate(operation: str, num1: float, num2: float) -> float:
```
Comment (hectordacb) [inline solutions/basic_calculator/main.py]: The function must match the main.py  name 🧑‍💻
[ ] The function's name matches the file name

**Review by hectordacb — CHANGES_REQUESTED** [review]
Hi @AnaiMurillo, the code is looking fantastic. There are just a few minor adjustments needed 😊

File: solutions/basic_calculator/basic_calculator.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+Module contents:
+    - calculate: Performs the specified arithmetic operation on two numbers
+
+Created on 28/12/2024
+Author: Ana Isabel Murillo
+"""
+
+
+def calculate(operation: str, num1: float, num2: float) -> float:
```
Comment (hectordacb) [inline solutions/basic_calculator/basic_calculator.py]: [ ] The function's name matches the file name

**Review by hectordacb — CHANGES_REQUESTED** [review]
The code looks good, just requires a small change

**Review by hectordacb — APPROVED** [review]
The code successfully passed the entire checklist, which shows it was a job very well done. Everything is clear and works correctly. Congratulations on the great effort and excellent outcome!✅

File: solutions/basic_calculator/calculate.py
Comment (Mr-Glucose) [inline solutions/basic_calculator/calculate.py:1]: The name of the file could be more _specific_!

File: solutions/tests/test_basic_calculator.py
Comment (Mr-Glucose) [inline solutions/tests/test_basic_calculator.py:1]: **Wonderful** job!

**Review by Mr-Glucose — COMMENTED** [review]
**Fantastic job on this code!** 🎉  
Your attention to detail and the way you tackled this problem is truly impressive. The logic flows seamlessly, and it's clear you've put a lot of thought into making it both efficient and readable.  

Keep up the amazing work—you're doing great things, and your skills are only getting stronger from here. Keep shining! 💪✨
